File: src/net/csv/CsvWriter.cs
// FormatFactory.Csv — Standalone .NET CSV Target Writer Library
// Sprint: FORMAT-FACTORY-DOTNET-TARGET-WRITER-MWP-DOGFOOD-UNBLOCKING-001
// target_ff_library: FormatFactory.Csv.CsvWriter
// dogfood_status: IMPLEMENTED (FODS → CSV)
// commercial_product_ready: false — G11-G not approved

using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace FormatFactory.Csv;

/// <summary>
/// Standalone Format Factory target writer for CSV (comma-separated values).
///
/// Scope:
///   - Serializes rows of string fields to RFC 4180 compatible CSV.
///   - Quotes fields containing comma, double-quote, CR, or LF.
///   - Doubles internal double-quotes.
///   - Outputs UTF-8 without BOM, LF line endings.
///   - Supports both in-memory (WriteRows) and file (WriteRowsToFile) output.
///
/// Usage:
///   var csv = CsvWriter.WriteRows(rows);
///   CsvWriter.WriteRowsToFile(rows, "/path/output.csv");
///
/// MWP status: minimal viable product — production hardening is future work.
/// commercial_product_ready: false
/// </summary>
public class CsvWriter
{
    // -------------------------------------------------------------------------
    // Public API
    // -------------------------------------------------------------------------

    /// <summary>
    /// Serialize <paramref name="rows"/> and write to <paramref name="path"/> (static overload).
    /// Callable as <c>CsvWriter.WriteRows(rows, path)</c> or on an instance (static-on-instance).
    /// </summary>
    public static void WriteRows(IEnumerable<IEnumerable<string?>> rows, string path) => WriteRowsToFile(rows, path);

    /// <summary>Instance method: serialize a CsvDocument to a CSV string.</summary>
    public string Serialize(CsvDocument doc)
    {
        var allRows = new List<IEnumerable<string?>>();
        if (doc.Headers != null) allRows.Add(doc.Headers);
        allRows.AddRange(doc.Rows);
        return WriteRows(allRows);
    }

    /// <summary>Serialize <paramref name="rows"/> with the given header row prepended, to a CSV string.</summary>
    public static string WriteRows(IEnumerable<IEnumerable<string?>> rows, string[] headers)
    {
        var all = new List<IEnumerable<string?>> { headers };
        all.AddRange(rows);
        return WriteRows(all);
    }

    /// <summary>
    /// Serialize <paramref name="rows"/> to a CSV string.
    /// </summary>
    /// <param name="rows">Rows of fields. Null fields are treated as empty strings.</param>
    /// <returns>CSV content as a string with LF line endings.</returns>
    public static string WriteRows(IEnumerable<IEnumerable<string?>> rows)
    {
        ArgumentNullException.ThrowIfNull(rows);
        var sb = new StringBuilder();
        foreach (var row in rows)
        {
            var fields = new List<string>();
            foreach (var field in row)
                fields.Add(EscapeField(field));
            sb.Append(string.Join(",", fields));
            sb.Append('\n');
        }
        return sb.ToString();
    }

    /// <summary>
    /// Serialize <paramref name="rows"/> and write to <paramref name="path"/>.
    /// Creates parent directories as needed. UTF-8, no BOM.
    /// </summary>
    public static void WriteRowsToFile(IEnumerable<IEnumerable<string?>> rows, string path)
    {
        ArgumentNullException.ThrowIfNull(rows);
        if (string.IsNullOrWhiteSpace(path))
            throw new CsvWriterException("path must not be null or empty.");

        var dir = Path.GetDirectoryName(path);
        if (!string.IsNullOrEmpty(dir)) Directory.CreateDirectory(dir);

        var content = WriteRows(rows);
        // Normalize any \r\n to \n
        content = content.Replace("\r\n", "\n");
        File.WriteAllText(path, content, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false));
    }

    /// <summary>Serialize dictionary rows using headers as column order, returns CSV string.</summary>
    public static string WriteRows(List<Dictionary<string, string>> rows, List<string> headers)
    {
        ArgumentNullException.ThrowIfNull(rows);
        ArgumentNullException.ThrowIfNull(headers);
        var all = new List<IEnumerable<string?>> { headers };
        all.AddRange(rows.Select(r => (IEnumerable<string?>)headers.Select(h => r.TryGetValue(h, out var v) ? v : "")));
        return WriteRows(all);
    }

    /// <summary>Serialize dictionary rows using headers as column order, write to file.</summary>
    public static void WriteRowsToFile(List<Dictionary<string, string>> rows, List<string> headers, string path)
    {
        ArgumentNullException.ThrowIfNull(rows);
        ArgumentNullException.ThrowIfNull(headers);
        var all = new List<IEnumerable<string?>> { headers };
        all.AddRange(rows.Select(r => (IEnumerable<string?>)headers.Select(h => r.TryGetValue(h, out var v) ? v : "")));
        WriteRowsToFile(all, path);
    }

    /// <summary>Write rows (with optional header) to a stream as CSV (UTF-8, no BOM).</summary>
    public static void WriteToStream(IEnumerable<IEnumerable<string?>> rows, Stream stream, string[]? headers = null)
    {
        ArgumentNullException.ThrowIfNull(rows);
        ArgumentNullException.ThrowIfNull(stream);
        IEnumerable<IEnumerable<string?>> all = headers != null
            ? new List<IEnumerable<string?>> { headers }.Concat(rows)
            : rows;
        var bytes = System.Text.Encoding.UTF8.GetBytes(WriteRows(all));
        stream.Write(bytes, 0, bytes.Length);
    }

    /// <summary>Write string[][] rows (with header array) to a file.</summary>
    public static void WriteRowsToFile(IEnumerable<string[]> rows, string path, string[] headers)
    {
        var all = new List<IEnumerable<string?>> { headers };
        all.AddRange(rows);
        WriteRowsToFile(all, path);
    }

    /// <summary>Write string[][] rows (with header array) to a stream.</summary>
    public static void WriteToStream(IEnumerable<string[]> rows, Stream stream, string[] headers)
    {
        var all = new List<IEnumerable<string?>> { headers };
        all.AddRange(rows);
        WriteToStream(all, stream);
    }

    /// <summary>Write a CsvDocument to a stream as CSV (UTF-8, no BOM).</summary>
    public static void WriteToStream(CsvDocument doc, Stream stream)
    {
        ArgumentNullException.ThrowIfNull(doc);
        ArgumentNullException.ThrowIfNull(stream);
        var bytes = System.Text.Encoding.UTF8.GetBytes(doc.ToCsv());
        stream.Write(bytes, 0, bytes.Length);
    }

    // -------------------------------------------------------------------------
    // Field escaping — RFC 4180 compatible
    // -------------------------------------------------------------------------

    /// <summary>
    /// Escape a single CSV field:
    /// - null/empty → empty string (no quotes)
    /// - Contains comma, double-quote, CR, or LF → wrap in double-quotes
    /// - Embedded double-quotes → doubled ("")
    /// </summary>
    public static string EscapeField(string? value)
    {
        if (value is null || value.Length == 0) return string.Empty;

        bool needsQuoting = value.IndexOf(',') >= 0 ||
                            value.IndexOf('"') >= 0 ||
                            value.IndexOf('\n') >= 0 ||
                            value.IndexOf('\r') >= 0;

        if (!needsQuoting) return value;
        return "\"" + value.Replace("\"", "\"\"") + "\"";
    }
}

/// <summary>Thrown by <see cref="CsvWriter"/> when output cannot be written.</summary>
public sealed class CsvWriterException : Exception
{
    public CsvWriterException(string message) : base(message) { }
    public CsvWriterException(string message, Exception inner) : base(message, inner) { }
}

File: src/python/csv/csv_writer.py
"""
csv_writer.py — CSV writer for format-factory-csv.

Public API:
  write_csv(rows, headers=None)  — returns CSV string (RFC 4180)
  write_csv_to_file(rows, path, headers=None) — writes CSV to file

NOTE: Does NOT import the stdlib 'csv' module (same reason as csv_parser.py).

Sprint: MAINSTREAM-MEGATRAIN-20260610
License: Apache-2.0
"""

from __future__ import annotations

from pathlib import Path
from typing import Any


class CsvWriteError(Exception):
    """Raised when CSV output fails."""


def _escape_field(value: str | None) -> str:
    """Escape a single CSV field per RFC 4180."""
    if value is None:
        return ""
    s = str(value)
    if not s:
        return ""
    needs_quoting = any(c in s for c in (",", '"', "\n", "\r"))
    if needs_quoting:
        return '"' + s.replace('"', '""') + '"'
    return s


def write_csv(
    rows: list[list[str | None]],
    headers: list[str] | None = None,
    delimiter: str = ",",
) -> str:
    """Serialize rows to RFC 4180 CSV string.

    Args:
        rows: Data rows, each a list of field values.
        headers: Optional header row prepended to output.
        delimiter: Field delimiter (default comma).

    Returns:
        CSV string with LF line endings.
    """
    if rows is None:
        raise CsvWriteError("rows must not be None")

    lines: list[str] = []
    if headers is not None:
        lines.append(delimiter.join(_escape_field(h) for h in headers))

    for row in rows:
        lines.append(delimiter.join(_escape_field(f) for f in row))

    return "\n".join(lines) + "\n" if lines else ""


def write_csv_to_file(
    rows: list[list[str | None]],
    path: str | Path,
    headers: list[str] | None = None,
    delimiter: str = ",",
) -> None:
    """Write CSV rows to a file. Creates parent dirs as needed. UTF-8, no BOM."""
    if not path:
        raise CsvWriteError("path must not be empty")
    p = Path(path)
    p.parent.mkdir(parents=True, exist_ok=True)
    content = write_csv(rows, headers=headers, delimiter=delimiter)
    p.write_text(content, encoding="utf-8")


def parse_and_rewrite(
    input_path: str | Path,
    output_path: str | Path,
    transform: Any = None,
) -> dict[str, Any]:
    """Roundtrip: read CSV, optionally transform, write back.

    Args:
        input_path: Source CSV file.
        output_path: Destination CSV file.
        transform: Optional callable(rows, headers) -> (rows, headers).

    Returns:
        Dict with row_count, column_count.
    """
    from .csv_parser import parse_csv_strict

    result = parse_csv_strict(input_path)
    headers = result.get("headers")
    rows = result.get("rows", [])

    if transform is not None:
        rows, headers = transform(rows, headers)

    write_csv_to_file(rows, output_path, headers=headers, delimiter=result.get("delimiter", ","))

    return {
        "row_count": len(rows),
        "column_count": len(headers) if headers else (len(rows[0]) if rows else 0),
    }
